gen/zap.go

```diff
@@ -0,0 +1,132 @@
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"fmt"
+
+	"go.uber.org/thriftrw/compile"
+)
+
+type zapGenerator struct {
+	mapG  mapGenerator
+	setG  setGenerator
+	listG listGenerator
+}
+
+func (z *zapGenerator) zapEncoder(g Generator, spec compile.TypeSpec) string {
+	root := compile.RootTypeSpec(spec)
+
+	switch t := root.(type) {
+	// Primitives
+	case *compile.BoolSpec:
+		return "Bool"
+	case *compile.I8Spec:
+		return "Int8"
+	case *compile.I16Spec:
+		return "Int16"
+	case *compile.I32Spec:
+		return "Int32"
+	case *compile.I64Spec:
+		return "Int64"
+	case *compile.DoubleSpec:
+		return "Float64"
+	case *compile.StringSpec:
+		return "String"
+	case *compile.BinarySpec:
+		return "String"
+
+	// Containers
+	case *compile.MapSpec:
+		switch t.KeySpec.(type) {
+		case *compile.StringSpec:
+			return "Object"
+		default:
+			return "Array"
+		}
+	case *compile.SetSpec, *compile.ListSpec:
+		return "Array"
+
+	// User-defined
+	case *compile.EnumSpec, *compile.StructSpec:
+		return "Object"
+	}
+	panic(root)
+}
+
+func (z *zapGenerator) zapMarshaler(g Generator, spec compile.TypeSpec, fieldValue string) (string, error) {
+	root := compile.RootTypeSpec(spec)
+
+	if _, ok := spec.(*compile.TypedefSpec); ok {
+		// For typedefs, cast to the root type and rely on that functionality.
+		rootName, err := typeReference(g, root)
+		if err != nil {
+			return "", err
+		}
+		fieldValue = fmt.Sprintf("(%v)(%v)", rootName, fieldValue)
+	}
+
+	if isPrimitiveType(spec) {
+		return fieldValue, nil
+	}
+
+	switch t := root.(type) {
+	case *compile.BinarySpec:
+		// There is no AppendBinary for ArrayEncoder, so we opt for encoding it ourselves and
+		// appending it as a string. We also use AddString instead of AddBinary for ObjectEncoder
+		// for consistency.
+		base64 := g.Import("encoding/base64")
+		return fmt.Sprintf("%v.StdEncoding.EncodeToString(%v)", base64, fieldValue), nil
+	case *compile.MapSpec:
+		return z.mapG.zapMarshaler(g, spec, t, fieldValue)
+	case *compile.SetSpec:
+		return z.setG.zapMarshaler(g, spec, t, fieldValue)
+	case *compile.ListSpec:
+		return z.listG.zapMarshaler(g, spec, t, fieldValue)
+	case *compile.StructSpec:
+		return fieldValue, nil
+	}
+	panic(root)
+}
+
+func (z *zapGenerator) zapMarshalerPtr(g Generator, spec compile.TypeSpec, fieldValue string) (string, error) {
+	if isPrimitiveType(spec) {
+		fieldValue = "*" + fieldValue
+	}
+	return z.zapMarshaler(g, spec, fieldValue)
+}
+
+func (z *zapGenerator) zapCanError(spec compile.TypeSpec) bool {
+	root := compile.RootTypeSpec(spec)
+
+	switch root.(type) {
+	// Primitives
+	case *compile.BoolSpec, *compile.I8Spec, *compile.I16Spec, *compile.I32Spec,
+		*compile.I64Spec, *compile.DoubleSpec, *compile.StringSpec, *compile.BinarySpec:
+		return false
+
+	// Non-primitives
+	case *compile.MapSpec, *compile.SetSpec, *compile.ListSpec, *compile.EnumSpec,
+		*compile.StructSpec:
+		return true
+	}
+	panic(root)
```

Comment:
Let's drop the panic. You can add checks for the true cases since there are
fewer of those and return false by default for all other cases.

Alternatively, you can also just call `zapEncoder` on the type and return true
only if the output is "Object" or "Array".

(If you do the zapEncodeBegin/zapEncodeEnd business, you can probably drop
this whole function.)

Reply:
did you want to drop panic for the other functions as well?

Reply:
For the cases we're doing exhaustive checks, let's keep the panic (but with a better error message).

For cases like this where it's more of an explicit list of things that can error out, we can just do catch-all for the false case rather than panicking.